pkg/cmd/service/service_test.go

```diff
@@ -20,118 +25,158 @@ import (
+func genCert(template, parent *x509.Certificate, pub *rsa.PublicKey, priv *rsa.PrivateKey) (*x509.Certificate, []byte, error) {
+	certBytes, err := x509.CreateCertificate(rand.Reader, template, parent, pub, priv)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cert, err := x509.ParseCertificate(certBytes)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	block := &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: certBytes,
+	}
+
+	return cert, pem.EncodeToMemory(block), nil
+}
+
+func genCACert() (*x509.Certificate, []byte, *rsa.PrivateKey, error) {
+	priv, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	var rootTemplate = &x509.Certificate{
+		SerialNumber:          big.NewInt(1),
+		KeyUsage:              x509.KeyUsageCertSign | x509.KeyUsageCRLSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+		MaxPathLen:            2,
+		NotBefore:             time.Now().Add(-time.Minute),
+		NotAfter:              time.Now().Add(time.Hour),
+		Subject: pkix.Name{
+			Country:      []string{"Earth"},
+			Organization: []string{"Starfleet"},
+		},
+		DNSNames: []string{"localhost"},
+	}
+
+	rootCert, rootPEM, err := genCert(rootTemplate, rootTemplate, &priv.PublicKey, priv)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	return rootCert, rootPEM, priv, nil
+}
+
+func genServerCert(caCert *x509.Certificate, caKey *rsa.PrivateKey) (*x509.Certificate, []byte, *rsa.PrivateKey, error) {
+	priv, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	var template = &x509.Certificate{
+		SerialNumber:          big.NewInt(1),
+		KeyUsage:              x509.KeyUsageCRLSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		IsCA:                  false,
+		NotBefore:             time.Now().Add(-time.Minute),
+		NotAfter:              time.Now().Add(time.Hour),
+		Subject: pkix.Name{
+			Country:      []string{"Earth"},
+			Organization: []string{"Starfleet"},
+		},
+		DNSNames: []string{"localhost"},
+	}
+
+	serverCert, serverPEM, err := genCert(template, caCert, &priv.PublicKey, caKey)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	return serverCert, serverPEM, priv, nil
+}
+
+func writeToTempFile(data []byte) (*os.File, error) {
+	file, err := os.CreateTemp("", "openfga_tls_test")
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = file.Write(data)
+	if err != nil {
+		return nil, err
+	}
 
-	_, err = certFile.Write([]byte(serverCert))
-	require.NoError(t, err, "failed to write pem")
+	return file, nil
+}
 
-	keyFile, err := ioutil.TempFile("", "server_key")
-	require.NoError(t, err, "error creating pem file")
-	defer keyFile.Close()
+type certChain struct {
+	caCert         *x509.Certificate
+	serverCertFile string
+	serverKeyFile  string
+}
+
+func (c certChain) Clean() {
+	os.Remove(c.serverKeyFile)
```

Comment:
Did you mean `c.serverCertFile` in this one? This is a duplicate of the one below.